RavenDB-23453 VectorSearch filter - RavenDB part

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23453
### Additional description

Implements usage of filter VectorSearch API introduced in #21898.
Introduces:
- CoraxConfigurationTest: Now allows injecting custom configuration into the IndexSearcher created by the query endpoint.
- Configuration options: Added a threshold for the filtered vector search scanning mode, configurable individually per index.
 - CoraxQueryBuilder: Refactored the query planner to accommodate filter requirements. It now rebalances the `AND` AST (where possible) to structure the query as `(VectorSearch AND (rest of clauses)).` This allows the other clauses to be injected as filters for the vector search.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [x] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing
- [x] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-23453 Moves GrowableBitArray into Sparrow project.

ayende commented on src/Corax/Querying/Matches/MultiVectorSearchMatch.cs:
```diff
@@ -14,35 +14,37 @@ public struct MultiVectorSearchMatch : IQueryMatch
     private bool _persisted = false;
     private int _positionOnPersistedValues = 0;
-
+    private IQueryMatch _filterQuery;
```
I don't think that this is what we want to do.
It isn't currently used, but isn't this basically about the same thing as evaluating each match directly here?
That can be expensive, no?

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -115,8 +132,26 @@ private void CreateDistanceBuffer(int length)
     }
 
-    private void FillAndPersistResults()
+    private void LoadFilterIfNeeded(Span<long> workingBuffer)
```
I don't understand what is going on here.
If the filter finds 1M results?

maciejaszyk committed:
RavenDB-23453 Adjusted tests and API of classes required for VectorSearch filter.

maciejaszyk committed:
RavenDB-23453 Filtered Vector Search: implement filtering inside VectorSearch.

Modifications:
- Scanning mode: translate document IDs from the filter into nodes and perform ExactSearch on them.
- ExactSearch: when a filter is provided, keep all node statistics so we can return results from best to worst that fulfill the filter conditions.
- NearestSearch: leverage the stateful (cached) SearchState and rerun the query with a larger NumberOfCandidates to hopefully find nodes that are accepted by the filter conditions.

maciejaszyk committed:
RavendB-23453 GrowableBitArray: changed iterator into struct to avoid unnecessary allocations.

maciejaszyk committed:
RavenDB-23453
- Assign `QueryPlan` only after at least one `Fill` call.
- Reuse already calculated distances in `VectorSearchRetriever`.

maciejaszyk committed:
RavenDB-23453 Added benchmark for vector search (without filter).

maciejaszyk committed:
RavenDB-23453 Increase NumberOfCandidates for NearestSearch due to filter internally.

maciejaszyk committed:
RavenDB-23453
- Dispose of allocations made by SearchState
- Benchmark based on actual embeddings
- NativeList: Reset Count to 0 upon disposal
- PR comments

maciejaszyk committed:
RavenDB-23453: Refactor Corax vector search and query building

- VectorSearch filter
- Introduced CoraxVectorItem as a CoraxQueryBuilder object for vector search
- Optimized AndQuery construction by separating VectorSearch from other clauses during the filter building phase
- Added CoraxTestingConfiguration to allow injecting testing configurations into IndexSearcher
- Updated tests to match the new API
- Refactored CoraxBooleanQueryBase and derived classes to improve readability and simplicity

maciejaszyk committed:
RavenDB-23453 Handle reversed order of AndNot expression.

maciejaszyk commented:
@ravendb-bot test this in debug please

maciejaszyk committed:
RavenDB-23453 Include epsilon to distance comparison in tests.

maciejaszyk committed:
RavenDB-23453
- VectorSearchMatch is empty when we have not found any nodes to scan.
- Addressed PR comments.

maciejaszyk committed:
RavenDB-23453 Rebase fixes

maciejaszyk committed:
RavenDB-23453
- Added `ShouldContinueSearch` to manage search termination based on filter document count.
- Improved candidate prefetch size calculation for `NearestSearcher`.

maciejaszyk committed:
RavenDB-23453
- Stop condition in post-filter VectorSearch is based on I/O
- Fix QueryPlan when filter is empty (vector search is not be initialized in such scenario).

maciejaszyk committed:
RavenDB-23453 Initial nodes in approximate search come from a filter.

maciejaszyk committed:
RavenDB-23453 Initial nodes in approximate search come from a filter.

maciejaszyk committed:
RavenDB-23453
- Introduced `Random` support for filtered vector search to enable deterministic testing.
- Adjusted `ShouldContinueSearch` logic for small filter sets.

maciejaszyk committed:
RavenDB-23453 Exact search: Cache distance calculations and optimize I/O operations.

maciejaszyk commented:
Todo:
Adjust changes from Federico's PR into this PR.
https://github.com/ravendb/ravendb/pull/22265

arekpalinski commented on src/Corax/Querying/Matches/MultiVectorSearchMatch.cs:
```diff
@@ -1,137 +1,223 @@
 public struct MultiVectorSearchMatch : IQueryMatch
 {
-    private readonly IndexSearcher _searcher;
+    private const int ScanningThreshold = 1024;
```
It was supposed to be configurable, wasn't it?

As per comment: https://github.com/ravendb/ravendb/pull/21898#discussion_r2597585179

> Yes. This is going to be configurable in the PR related to RavenDB changes.

arekpalinski commented on src/Corax/Querying/IndexSearcher.cs:
```diff
@@ -30,6 +30,12 @@ namespace Corax.Querying;
+    /// Used for testing purposes only.
+    /// </summary>
+    private readonly CoraxTestingConfiguration _testingConfiguration;
+    internal bool VectorSearchScanningDisabled => _testingConfiguration is not null && _testingConfiguration.DisableVectorSearchScanning;
```
Can we rename it to something like `VectorSearchScanningDisabledForTesting` as it is used only for testing purposes?

maciejaszyk replied:
I've removed it entirely. It is not necessary to have an explicit bool flag. We can check it directly when needed.

arekpalinski commented on src/Corax/Querying/IndexSearcher.cs:
```diff
@@ -141,8 +147,9 @@ private void Init()
     }
 
-    private IndexSearcher(IndexFieldsMapping fieldsMapping)
+    private IndexSearcher(IndexFieldsMapping fieldsMapping, CoraxTestingConfiguration testingConfiguration)
```
Do we really need to pass it via constructor?

maciejaszyk replied:
Removed

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:
```diff
@@ -647,6 +530,30 @@ IQueryMatch HandleStringUnaryMatch(Parameters queryEnvironment)
+
+        // We internally negated the right expression. If we find a pattern true and (NOT EXPR) we can skip the true, since it's noop. 
+        if (leftExpr is TrueExpression)
+            return right; // true and not... optimization
```
Does it conflict somehow with https://github.com/ravendb/ravendb/pull/22320 or needs to be adjusted?

redknightlois replied:
Yes, the PR is doing the same thing but handling also the recursive cases.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -5626,6 +5627,8 @@ internal TestingStuff ForTestingPurposesOnly()
 
         internal sealed class TestingStuff
         {
+            internal CoraxTestingConfiguration CoraxConfiguration;
```
I wonder if we should create `TestingStuff` in Corax's `IndexSearcher.cs` and put `CoraxTestingConfiguration` there.

redknightlois replied:
Agree centralizing testing facilities is important.

maciejaszyk replied:
CoraxTestingConfiguration is actually from the Corax project. However, this field is made explicitly for testing Corax integration inside RavenDB. For example, to pass custom configuration during IndexReadOperation creation. There is no other way to inject it from test.

redknightlois commented on src/Corax/Querying/Matches/MultiVectorSearchMatch.cs:
```diff
@@ -1,137 +1,223 @@
+                {
+                    true => vectorSearcher.Fill(matchBuffer, distanceBuffer),
+                    false => vectorSearcher.Fill(matchBuffer, distanceBuffer, _filterResults),
+                };
```
Isn't this better with something like:
```
var fillResults = _filterQuery is null ? null : _filterResults;
vectorSearch.Fill(matchBuffer, distanceBuffer, fillResults);
```

maciejaszyk replied:
done

redknightlois commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -129,94 +228,35 @@ private void FillAndPersistResults()
+                true => _vectorSearchRetriever.Fill(mBuf, dBuf),
+                false => _vectorSearchRetriever.Fill(mBuf, dBuf, _filterResults)
+            };
+            
```
Same as in https://github.com/ravendb/ravendb/pull/21737/changes#r2828423806

maciejaszyk replied:
done

redknightlois commented on src/Corax/Querying/IndexSearcher.cs:
```diff
@@ -49,7 +55,7 @@ public sealed unsafe partial class IndexSearcher : IDisposable
     public bool ForceNonAccelerated { get; set; }
 
-    public bool IsAccelerated => AdvInstructionSet.IsAcceleratedVector256 && !ForceNonAccelerated;
+    public bool IsAccelerated => AdvInstructionSet.IsAcceleratedVector256 && (_testingConfiguration?.IsAccelerated ?? ForceNonAccelerated == false);
```
Cant we just use the `AdvInstructionSet` tests capability to do that instead?

redknightlois replied:
Also unless I am reading it wrong, when any `CoraxTestingConfiguration` instance is provided (even with default `IsAccelerated = true`), the null-coalescing short-circuits and `ForceNonAccelerated` is never checked. A searcher with `ForceNonAccelerated = true` and a `CoraxTestingConfiguration()` will still report `IsAccelerated = true` on  AVX2-capable hardware.

maciejaszyk replied:
Actually, I've removed ForceNonAccelerated since we're not using it in production code—only for testing. Nonetheless, the way to disable it now (explicitly in tests) is via the testing configuration. This is more consistent, IMHO.

redknightlois commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
This is unreachable now.

maciejaszyk replied:
yes, removed

redknightlois commented on src/Corax/Querying/IndexSearcher.VectorSearch.cs:
```diff
@@ -6,65 +6,164 @@
+
+            var uniqueCount = Sorting.SortAndRemoveDuplicates(nodesIdsToScan.ToSpan());
+            nodesIdsToScan.Count = uniqueCount;
+            return nodesIdsToScan.Count > 0;
```
If `nodesIdsToScan.Count` is 0, the `nodesIdsToScan` get's allocated and since we are returning `false` we are assuming the parent will not `.Dispose()` this can cause a memory leak in the context which while at the end of the request it will be reclaimed, I don't know how many times this condition could happen inflating memory usage on potentially complex scenarios.

maciejaszyk replied:
Good catch!

maciejaszyk committed:
RavenDB-23453 Refactor HNSW search logic

maciejaszyk committed:
RavenDB-23453  Implement RandomNodesFromFilterEnumerator for efficient random node selection from filter results

maciejaszyk committed:
RavenDB-23453  Implement RandomNodesFromFilterEnumerator for efficient random node selection from filter results

maciejaszyk committed:
RavenDB-23453 Refactor vector search methods to accept optional filter parameter and update related tests
